load config files in order, later overrides earlier, lots more testing

jaidhyani opened the pull request:
The main functional impact of this PR is to kill the "priority" config arg and just load configs in the order in which they're passed in to the training script, with the later overriding the former. Alongside this we do some refactoring to support testing to ensure that everything works as expected. We also improve type-casting override argument strings to better match the actual config types and make it significantly less dangerous.

jettjaniak commented:
why the merge commit?

jaidhyani commented:
> why the merge commit?

Default behavior in the github UI, clicked without thinking. I think we can configure this to default to rebase/ff.

<img width="937" alt="image" src="https://github.com/delphi-suite/delphi/assets/381155/0dc7f3b0-f507-4fbb-bfbf-2b51898a496e">

jettjaniak commented:
https://github.com/orgs/community/discussions/12032
not configurable unfortunately :/

jettjaniak commented on .vscode/launch.json:
let's add it to gitignore

jettjaniak commented on scripts/run_training.py:
```diff
@@ -52,6 +55,8 @@ def setup_parser() -> argparse.ArgumentParser:
     parser = argparse.ArgumentParser(description="Train a delphi model")
     parser.add_argument(
         "--config_files",
+        "--config_file",
+        "-c",
```
```suggestion
        "--config",
        "-c",
```
I don't care about backward compatibility, no one is using this yet

jettjaniak commented on scripts/run_training.py:
```diff
@@ -78,20 +83,7 @@ def setup_parser() -> argparse.ArgumentParser:
 def overrides_to_dict(overrides: list[str]) -> dict[str, Any]:
     # {"--overrides a.b.c=4 foo=false} to {"a": {"b": {"c": 4}}, "foo": False}
```
that needs updating

jettjaniak replied:
I mean, what's an example input?

jettjaniak commented on src/delphi/constants.py:
```diff
@@ -1,7 +1,9 @@
 
 STATIC_ASSETS_DIR = files("delphi.static")
-CONFIG_PRESETS_DIR = STATIC_ASSETS_DIR / "configs"
+CONFIG_PRESETS_DIR = cast(Path, STATIC_ASSETS_DIR / "configs")
```
Is this a Path?

jaidhyani replied:
It is now. More generally, it is a Path in practice, and doing this casting here (1) breaks nothing and (2) makes some basic operations more straightforward with less boilerplate.

jettjaniak commented on src/delphi/train/config/utils.py:
```diff
@@ -1,10 +1,12 @@
+from typing import Any, Type, TypeVar, Union
 
 import platformdirs
 from beartype.typing import Any, Iterable
```
```suggestion
from collections.abc import Iterable
```

jettjaniak reviewed (approved):
nice! 🚀 
- why do we need `filter_config_to_actual_config_values`?
- in general I'm quite overwhelmed by train/config/utils.py, I didn't realize we'll need all of this to have the functionality I imagined
- could you add some docstrings? like what are the functions doing and why we need them

jaidhyani commented:
> * filter_config_to_actual_config_values

I think we actually don't need this anymore, going to optimistically delete.

jaidhyani commented:
> * 
> * in general I'm quite overwhelmed by train/config/utils.py, I didn't realize we'll need all of this to have the functionality I imagined
> * could you add some docstrings? like what are the functions doing and why we need them

I added some more docstrings explaining how some things are used, why we need them, and how they work. I also deleted a few things that we don't need anymore.
